### PR TITLE
fix(issuer): [SFEQS-1146] remove outdated fields when DOCUMENT status changes

### PR DESCRIPTION
**Pull request by lucacavallaro**
#### List of Changes

1.  `Document` is now a `tagged union` instead of a `flat structure`
2.  The state machine is now type-safe and now it returns the right fields associated for each `Document` status

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by grausof** [comment]
I have tested this code but it seems some things still don't work. In particular, on a document that was previously `REJECTED` and then becomes `READY`, the fields relating to rejected (`rejected_at`, `rejected_reason`) are not deleted from the DB. 
Screenshot 2022-11-22 alle 10 57 25

File: apps/io-func-sign-issuer/src/upload.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ import { Issuer } from "./issuer";
 export const UploadMetadata = t.intersection([
   t.type({
     id: Id,
-    documentId: Document.types[0].props.id,
+    documentId: DocumentId,
     signatureRequestId: SignatureRequest.types[0].props.id,
```
[the remark this answers is not in the document]
Comment (lucacavallaro) [reply apps/io-func-sign-issuer/src/upload.ts:20]: yes, only if the extraction of the decoder requires the access on `types[0]` property

**Comment by gunzip** [comment]
can we close this one or what? :-)

**Comment by grausof** [comment]
> can we close this one or what? :-)

The bug persists: https://github.com/pagopa/io-sign/pull/15#issuecomment-1323402513

**Comment by lucacavallaro** [comment]
@gunzip nope, this fix is ​​still in progress

**Review by gunzip — APPROVED** [review]

**Pull request merged**